stdlib/public/Reflection/Sources/_Runtime/ImageInspection.swift

```diff
@@ -0,0 +1,175 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+import _SwiftRuntimeShims
+
+#if !(os(macOS) || os(iOS) || os(tvOS) || os(watchOS))
+import SwiftShims
+#endif
+
+@available(SwiftStdlib 5.9, *)
+@frozen
+public enum ImageInspection {}
+
+//===----------------------------------------------------------------------===//
+// Mach-O Image Lookup
+//===----------------------------------------------------------------------===//
+
+#if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
+@_silgen_name("initializeDyldLookup")
+func initializeDyldLookup()
+#endif
+
+//===----------------------------------------------------------------------===//
+// ELF and COFF Image Lookup
+//===----------------------------------------------------------------------===//
+
+#if !(os(macOS) || os(iOS) || os(tvOS) || os(watchOS))
+func enumerateSections<K: Hashable, V>(
+  _ body: (MetadataSections) -> (
+      MetadataSectionRange,
+      (UnsafeRawPointer, Int) -> ()
+    )
+) {
+  swift_enumerateAllMetadataSections({ sections, _ in
+    guard let sections = sections else {
+      return true
+    }
+
+    let metadataSections = sections.assumingMemoryBound(
+      to: MetadataSections.self
+    )
+
+    guard metadataSections.version >= 3 else {
+      return false
+    }
+
+    let (range, register) = body(metadataSections)
+
+    register(range.start, range.length)
+
+    return true
+  }, nil)
+}
+#endif
+
+//===----------------------------------------------------------------------===//
+// Runtime Attributes
+//===----------------------------------------------------------------------===//
+
+@available(SwiftStdlib 5.9, *)
+struct AttributeCache {
+#if !(os(macOS) || os(iOS) || os(tvOS) || os(watchOS))
+  var imageCount = 0
+#endif
+  var map: [ContextDescriptor: [UnsafeRawPointer]] = [:]
+}
+
+@available(SwiftStdlib 5.9, *)
+var attributeCache: Lock<AttributeCache> = .create(with: .init())
+
+@available(SwiftStdlib 5.9, *)
+extension ImageInspection {
+  @available(SwiftStdlib 5.9, *)
+  public static func withAttributeCache<T>(
+    _ body: @Sendable ([ContextDescriptor: [UnsafeRawPointer]]) throws -> T
+  ) rethrows -> T {
+#if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
+    initializeDyldLookup()
+
+    return try attributeCache.withLock {
+      try body($0.map)
+    }
+#else
+    return try attributeCache.withLock {
+      // If our cached image count is less than what's being reported by the
+      // Swift runtime, then we need to reset our cache and enumerate all of
+      // the images again for updated attributes.
+      let currentImageCount = swift_getMetadataSectionCount()
```

Comment:
This call is unsafe in production—it can race with `enumerateSections()`. It should only be used for tests. :(